codyze-core/src/main/kotlin/de/fraunhofer/aisec/codyze/Sarif.kt

```diff
@@ -0,0 +1,216 @@
+
+    return if (this is FunctionDeclaration && this.body != null && onlyFunctionHeader) {
+            // Try to calculate the end of the header by using the beginning of the body. This is
+            // not entirely correct since in some programming languages we need to start the body
+            // location at the first statement, since we are missing location information for the
+            // body "block", but it's the best we can do
+            val location =
+                de.fraunhofer.aisec.cpg.sarif.PhysicalLocation(
+                    uri = location.artifactLocation.uri,
+                    region =
+                        de.fraunhofer.aisec.cpg.sarif.Region(
+                            startLine = location.region.startLine,
+                            startColumn = location.region.startColumn,
+                            endLine =
+                                this.body?.location?.region?.startLine ?: location.region.endLine,
+                            endColumn =
+                                this.body?.location?.region?.startColumn
+                                    ?: location.region.endColumn,
+                        ),
+                )
+            location
+        } else {
+            this.location
+        }
+        ?.let {
+            Location(
+                physicalLocation = it.toSarif(),
+                logicalLocations =
+                    listOf(
+                        LogicalLocation(
+                            fullyQualifiedName =
+                                if (this is Declaration) this.name.toString() else null,
+                            name = this.name.localName,
+                            kind = this.toSarifKind(),
+                        )
+                    ),
+                message = message,
+            )
+        }
+}
+
+/** Converts a [de.fraunhofer.aisec.cpg.sarif.PhysicalLocation] to a [PhysicalLocation]. */
+fun de.fraunhofer.aisec.cpg.sarif.PhysicalLocation.toSarif(): PhysicalLocation {
+    return PhysicalLocation(
+        artifactLocation = ArtifactLocation(uri = "file://${this.artifactLocation.uri.path}"),
+        region =
+            Region(
+                startLine = this.region.startLine.toLong(),
+                startColumn = this.region.startColumn.toLong(),
+                endLine = this.region.endLine.toLong(),
+                endColumn = this.region.endColumn.toLong(),
+            ),
+    )
+}
+
+/**
+ * Converts a [Node] to a well-known SARIF kind. This is used to categorize the node in the SARIF
+ * output.
+ */
+private fun Node.toSarifKind(): String? {
+    return when (this) {
+        is FunctionDeclaration -> "function"
+        is FieldDeclaration -> "member"
+        is NamespaceDeclaration -> "namespace"
+        is ParameterDeclaration -> "parameter"
+        is VariableDeclaration -> "variable"
+        is Declaration -> "declaration"
+        is Reference -> "object"
+        is Literal<*> -> "value"
```

Comment:
In SARIF "object" and "value" should be used within JSON documents, cf. https://docs.oasis-open.org/sarif/sarif/v2.1.0/errata01/os/sarif-v2.1.0-errata01-os-complete.html#_Toc141790976.

Reply:
Should a reference be a `variable` then?